### PR TITLE
Reform metering for value cloning and memory allocation

### PR DESCRIPTION
**Pull request by jayz22**
### What

Resolves https://github.com/stellar/rs-soroban-env/issues/604 and https://github.com/stellar/rs-soroban-env/issues/680.

This PR reforms `metered_clone`.
- Splits memory allocation from shallow copying
- Define the "shallow size" of `MeteredClone` types. 

A `metered_clone` consist of shallow copying of the type's memory footprint (for a `Vec` it's 24), plus recursively (deep) cloning any of the substructures. 
A `HostMemAlloc` charge occurs only when memory is allocated. I.e. types involving `alloc` -- `Box`, `Vec`, `Rc` etc.
The shallow copying of a type is charged the amount `ELT_SIZE` defined by the implementer, and there is a runtime check to ensure that amount is the upper bound of the actual memory size of the type `mem::size_of<T>`.
Adds a test in `budget_metering.rs` to demonstrate the metered cloning of a deep structure. 

### Why

[TODO: Why this change is being made. Include any context required to understand the why.]

### Known limitations

[TODO or N/A]

### CODE REVIEW

File: soroban-env-host/src/test/budget_metering.rs [hunk 1]
```diff
@@ -142,3 +144,51 @@ fn map_insert_key_vec_obj() -> Result<(), HostError> {
...
+    ];
+
+    v.charge_for_clone(host.as_budget())?;
+    // Transition #                  (1)           (2)             (3)
+    // Type(size, count):  Vec(24,1) ---> Box(8,3) ----> Vec(24,3) ------> ScMapEntry(80,3x2)
+    // MemAlloc:                     8x3           24x3            80x3x2                     = 576
+    // MemCpy:             24        8x3  8x3      24x3  24x3      80x3x2  80x3x2             = 1176
+    // At the alloc boundary there is some double-charging happening.
+    // For example in transition #1, `Vec` charges for (1) allocation of 8x3 bytes, (2) copying of
+    // 8x3 bytes data into it. The `Box` charges for 8x3 for its own size.
```
Comment (jayz22) [inline soroban-env-host/src/test/budget_metering.rs:180]: There is a bit of double counting going on. It comes from my current design and I’m not 100% sure if we can totally eliminate it: each type takes care itself (shallow copying of its own “pointer” size) and its substructures (allocating buffer, and shallow copying member “pointers” into the buffer). There isn’t really a way to know if itself is a substructure of something else, in which case its parent may have taken care of the shallow copying its pointer part. Thus leads to the shallow copy being double counted.

File: soroban-env-host/src/host/metered_clone.rs [hunk 2]
```diff
@@ -121,82 +185,256 @@ impl MeteredClone for ScVal {
...
+        }
+    }
+}
+
+mod test {
+    #[allow(unused)]
+    use super::*;
+
+    #[test]
+    fn test_elt_size() {
```
Comment (jayz22) [inline soroban-env-host/src/host/metered_clone.rs:357]: I'm not sure if there's a way to automatically cover this test on all types implementing the `MeteredClone` trait.

File: soroban-env-host/src/host/metered_clone.rs [hunk 3]
```diff
@@ -1,31 +1,45 @@
...
-// TODO: this is correct for CPU-cost but not memory-cost, since there's no heap
-// allocation implied by a shallow clone. We should (maybe) split BytesClone
-// into ByteCopy (the CPU byte-copying part) and HeapAlloc (the heap-allocating
-// part). See https://github.com/stellar/rs-soroban-env/issues/604.
-fn charge_shallow_copy(n_elts: usize, budget: &Budget) -> Result<(), HostError> {
-    budget.charge(CostType::BytesClone, n_elts as u64)
+extern crate static_assertions as sa;
+
+// Charge for an N-element "shallow copy" of some type, not cloning any substructure. The charge
+// unit is number of elements `n_elts` multiply by size of each element. In a better world we would
```
Comment (graydon) [inline soroban-env-host/src/host/metered_clone.rs:23]:
```suggestion
// unit is number of elements `n_elts` multiplied by a declared size of each element. In a better world we would
```

File: soroban-env-host/src/host/metered_clone.rs [hunk 4]
```diff
@@ -121,82 +185,256 @@ impl MeteredClone for ScVal {
...
-            // Shallow types we can batch-up our charging for.
-            budget.charge(CostType::BytesClone, self.len() as u64)?;
-            Ok(self.clone())
-        } else {
-            self.iter().map(|elt| elt.metered_clone(budget)).collect()
+
+    const ELT_SIZE: u64 = 24;
+
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        // first take care of Vec clone, which involves allocating memory, (shallowly) memcpy the
```
Comment (graydon) [inline soroban-env-host/src/host/metered_clone.rs:235]: Hm, this part I think isn't right -- it removed the special case of charging once for a clone of vectors of shallow elements.

Comment (graydon) [reply soroban-env-host/src/host/metered_clone.rs:235]: Except, wait, no, you _are_ charging for shallow copies of C, but then also charging for non-shallow cloning. I think I'm confused about what's going on here.

File: soroban-env-host/src/host/metered_clone.rs [hunk 5]
```diff
@@ -121,82 +185,256 @@ impl MeteredClone for ScVal {
...
     const IS_SHALLOW: bool = false;
 
-    fn metered_clone(&self, budget: &Budget) -> Result<Self, HostError> {
+    const ELT_SIZE: u64 = 8;
+
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        // first take care of the Box part
+        charge_shallow_copy::<Box<C>>(1, budget)?;
+        charge_heap_alloc::<C>(1, budget)?;
+        charge_shallow_copy::<C>(1, budget)?;
```
Comment (graydon) [inline soroban-env-host/src/host/metered_clone.rs:257]: Similarly this looks like it's charging for copying C .. twice? once shallow and once deep.

File: soroban-env-host/src/host/metered_clone.rs [hunk 6]
```diff
@@ -121,82 +185,256 @@ impl MeteredClone for ScVal {
...
+
+    const ELT_SIZE: u64 = 80;
+
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        charge_shallow_copy::<Self>(1, budget)?;
+        if self.args.is_heap() {
+            // equivalent to charging for a `Vec` clone.
+            charge_shallow_copy::<Vec<DebugArg>>(1, budget)?;
+            charge_heap_alloc::<Vec<DebugArg>>(self.args.len() as u64, budget)?;
+            charge_shallow_copy::<Vec<DebugArg>>(self.args.len() as u64, budget)
```
Comment (graydon) [inline soroban-env-host/src/host/metered_clone.rs:293]: another case of the shallow-type vector double charging situation here

File: soroban-env-host/src/host/metered_clone.rs [hunk 7]
```diff
@@ -35,14 +49,20 @@ pub trait MeteredClone: Clone {
     // `charge_for_clone` (correctly charging for cloning the substructure).
     const IS_SHALLOW: bool = true;
 
+    // Size (num of bytes) of a single element. This value determines the input for budget charging.
+    // It should be the upperbound (across various compilations and platforms) of the actual type's
+    // size. Implementer of the trait needs to decide this value based on Rust's guideline on type
+    // layout: https://doc.rust-lang.org/reference/type-layout.html
```
Comment (graydon) [inline soroban-env-host/src/host/metered_clone.rs:55]:
```suggestion
    // layout: https://doc.rust-lang.org/reference/type-layout.html
    //
    // Note: When `Self::IS_SHALLOW = false`, `ELT_SIZE` should only cover any _extra_ size of `Self` above
    // and beyond the size of its substructure. For example for `Vec<C>` we have `ELT_SIZE=24` to account
    // for the size of the vector header / fat-pointer part, _not_ the size of the elements. For a struct type
    // that is not shallow, typically `Self::ELT_SIZE` will be zero, or nearly zero (possibly covering any padding
    // that occurs between elements).
```

**Comment by jayz22** [comment]
@graydon I have reformed it once again after our discussion. All the previous double-charging of shallow copies should be eliminated. I believe this is ready for merge. Please take another look.

**Review by graydon — APPROVED** [review]

**Pull request merged**